File: src/church_day.rs
use chrono::NaiveDate;
use std::fmt;

#[derive(Copy, Clone)]
pub enum DayClass {
    Solemnity,
    Sunday,
    Feast,
    Memorial,
    SeasonalWeekday,
    FerialWeekday,
    DayOfFastingAndAbstinenceFromDefinition,
}

impl fmt::Display for DayClass {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            DayClass::Solemnity => write!(f, "Solemnity"),
            DayClass::Sunday => write!(f, "Sunday"),
            DayClass::Feast => write!(f, "Feast"),
            DayClass::Memorial => write!(f, "Memorial"),
            DayClass::SeasonalWeekday => write!(f, "Seasonal Weekday"),
            DayClass::FerialWeekday => write!(f, "Ferial"),
            DayClass::DayOfFastingAndAbstinenceFromDefinition => {
                write!(f, "Day of Fasting and Abstinence")
            }
        }
    }
}

#[derive(Copy, Clone)]
pub enum Season {
    Advent,
    Christmas,
    OrdinaryTime,
    Lent,
    Triduum,
    Easter,
}

impl fmt::Display for Season {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            Season::Advent => write!(f, "Advent"),
            Season::Christmas => write!(f, "Christmas"),
            Season::OrdinaryTime => write!(f, "Ordinary Time"),
            Season::Lent => write!(f, "Lent"),
            Season::Triduum => write!(f, "Sacred Triduum"),
            Season::Easter => write!(f, "Easter"),
        }
    }
}

#[derive(Copy, Clone)]
pub struct ChurchDay {
    pub date: NaiveDate,
    pub class: DayClass,
    pub season: Season,
}

impl ChurchDay {
    pub fn new(year: i32, month: u32, day: u32, class: DayClass, season: Season) -> Self {
        Self {
            date: NaiveDate::from_ymd_opt(year, month, day)
                .expect("Unable to instantiate new ChurchDay"),
            class,
            season,
        }
    }

    pub fn month_name(&self) -> String {
        self.date.format("%B").to_string()
    }

    pub fn string_day_of_week(&self) -> String {
        self.date.format("%A").to_string()
    }
}

pub struct ChurchYear {
    pub days: Vec<ChurchDay>,
}

File: src/church_day.c
#include "church_day.h"
#include <string.h>
#include <stdlib.h>
#include <stdio.h>

ChurchDay church_day_new(int year, int month, int day, DayClass class, Season season, const char *description) {
	ChurchDay cd;
	memset(&cd.date, 0, sizeof(struct tm));

	cd.date.tm_year = year - 1900;
	cd.date.tm_mon = month - 1;
	cd.date.tm_mday = day;

	mktime(&cd.date);

	cd.class = class;
	cd.season = season;
	strncpy(cd.description, description, sizeof(cd.description) - 1);
	cd.description[sizeof(cd.description) - 1] = '\0';

	return cd;
}

const char* church_day_month_name(const ChurchDay *day) {
	static char buffer[32];
	strftime(buffer, sizeof(buffer), "%B", &day->date);
	return buffer;
}

const char* church_day_weekday_string(const ChurchDay *day) {
	static char buffer[32];
	strftime(buffer, sizeof(buffer), "%A", &day->date);
	return buffer;
}

const char* day_class_to_string(DayClass class) {
	switch (class) {
		case SOLEMNITY: return "Solemnity";
		case SUNDAY: return "Sunday";
		case FEAST: return "Feast";
		case MEMORIAL: return "Memorial";
		case SEASONAL_WEEKDAY: return "Seasonal Weekday";
		case FERIAL_WEEKDAY: return "Ferial";
		default: return "Unknown";
	}
}

const char* season_to_string(Season season) {
	switch(season) {
		case ADVENT: return "Advent";
		case CHRISTMAS: return "Christmas";
		case ORDINARY_TIME: return "Ordinary Time";
		case LENT: return "Lent";
		case TRIDUUM: return "Sacred Triduum";
		case EASTER: return "Easter";
		default: return "Unknown";
	}
}

ChurchYear church_year_new(void) {
	ChurchYear cy;
	cy.capacity = 366;
	cy.count = 0;
	cy.days = malloc(cy.capacity * sizeof(ChurchDay));
	return cy;
}

void church_year_add(ChurchYear *year, ChurchDay day) {
	if (year->count >= year->capacity) {
		year->capacity *= 2;
		year->days = realloc(year->days, year->capacity * sizeof(ChurchDay));
	}
	year->days[year->count++] = day;
}

void church_year_free(ChurchYear *year) {
	free(year->days);
	year->days = NULL;
	year->count = 0;
	year->capacity = 0;
}
